ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java

```diff
@@ -406,21 +412,70 @@ private boolean containsHostFromHostGroups(String configType, String propertyNam
         clusterConfig.setProperty("hdfs-site", "dfs.internal.nameservices", nameservices);
       }
 
-      // if the active/stanbdy namenodes are not specified, assign them automatically
-      if (! isNameNodeHAInitialActiveNodeSet(clusterProps) && ! isNameNodeHAInitialStandbyNodeSet(clusterProps)) {
-        Collection<String> nnHosts = clusterTopology.getHostAssignmentsForComponent("NAMENODE");
-        if (nnHosts.size() < 2) {
-          throw new ConfigurationTopologyException("NAMENODE HA requires at least 2 hosts running NAMENODE but there are: " +
-              nnHosts.size() + " Hosts: " + nnHosts);
+      // parse out the nameservices value
+      String[] parsedNameServices = parseNameServices(hdfsSiteConfig);
+
+      // if a single nameservice is configured (default HDFS HA deployment)
+      if (parsedNameServices.length == 1) {
+        LOG.info("Processing a single HDFS NameService, which indicates a default HDFS NameNode HA deployment");
+        // if the active/stanbdy namenodes are not specified, assign them automatically
+        if (! isNameNodeHAInitialActiveNodeSet(clusterProps) && ! isNameNodeHAInitialStandbyNodeSet(clusterProps)) {
+          Collection<String> nnHosts = clusterTopology.getHostAssignmentsForComponent("NAMENODE");
+          if (nnHosts.size() < 2) {
+            throw new ConfigurationTopologyException("NAMENODE HA requires at least 2 hosts running NAMENODE but there are: " +
+                nnHosts.size() + " Hosts: " + nnHosts);
+          }
+
+          // set the properties that configure which namenode is active,
+          // and which is a standby node in this HA deployment
+          Iterator<String> nnHostIterator = nnHosts.iterator();
+          clusterConfig.setProperty("hadoop-env", "dfs_ha_initial_namenode_active", nnHostIterator.next());
+          clusterConfig.setProperty("hadoop-env", "dfs_ha_initial_namenode_standby", nnHostIterator.next());
+
+          configTypesUpdated.add("hadoop-env");
         }
+      } else {
+        if (!isPropertySet(clusterProps, "hadoop-env", HDFS_ACTIVE_NAMENODE_SET_PROPERTY_NAME) && !isPropertySet(clusterProps, "hadoop-env", HDFS_STANDBY_NAMENODE_SET_PROPERTY_NAME)) {
+          // multiple nameservices indicates an HDFS NameNode Federation install
+          // process each nameservice to determine the active/standby nodes
+          LOG.info("Processing multiple HDFS NameService instances, which indicates a NameNode Federation deployment");
+          if (parsedNameServices.length > 1) {
+            Set<String> activeNameNodeHostnames = new HashSet<String>();
+            Set<String> standbyNameNodeHostnames = new HashSet<String>();
+
+            for (String nameService : parsedNameServices) {
+              List<String> hostNames = new LinkedList<String>();
```

Comment:
`hostNames` is accessed by index below, which does not scale well for `LinkedList`.  Either use `ArrayList` or access by iterator.

(Not really a performance problem here, since the list is small, but good to note in general.)

Reply:
Changed to ArrayList.  Thanks for mentioning this.